src/Workspaces/Core/Portable/FindSymbols/Shared/AbstractSyntaxIndex.cs

```diff
@@ -103,11 +103,28 @@ protected static async ValueTask<TIndex> GetRequiredIndexAsync(Document document
         }
 
         private static async Task<TIndex> CreateIndexAsync(
-            Document document, Checksum checksum, IndexCreator create, CancellationToken cancellationToken)
+            Document document,
+            Checksum textChecksum,
+            Checksum textAndDirectivesChecksum,
+            IndexCreator create,
+            CancellationToken cancellationToken)
         {
             Contract.ThrowIfFalse(document.SupportsSyntaxTree);
 
             var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+
+            // if the tree contains directives, then include the directives-checksum info in the checksum we produce. We
+            // don't want to consider the data reusable if the user changes pp directives.  Note: currently, this
+            // unfortunately means we cannot reuse indices created for documents that use `#nullable` or `#region`.
+            // Ideally, this would only impact documents containing `#if` directives.
+            //
+            // It's trivial for us to determine the checksum to use at the index-creation/writing point because we have
+            // to have computed the syntax tree anyways to produce the index.  The tradeoff of this design though is
+            // that at the reading point we may have to issue two reads to determine which case we're in.  However, this
+            // still let's us avoid parsing the doc at the point we're reading in the indices (which would defeat a
+            // major reason for having the index in the first place).
+            var checksum = root.ContainsDirectives ? textAndDirectivesChecksum : textChecksum;
```

Comment:
i will likely optimize this further to actually walk the tree to find out if which directive types these are.  walking a tree is *super* cheap since we can bail out of any nodes that don't have directives.